### PR TITLE
feat: Tails for Pages

### PR DESCRIPTION
**Pull request by jrf0110**
Adds the `wrangler pages deployment tail [deploymentId]` command

### CODE REVIEW

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+): data is StatusChoice[] =>
+	data?.every((d) => statusChoices.includes(d)) ?? false;
+
+export function Options(yargs: Argv) {
+	return (
+		yargs
+			.positional("deployment-id", {
+				type: "string",
+				description:
+					"Deployment to tail. Specify by environment if " +
```
Comment (jrf0110) [inline packages/wrangler/src/pages/deployment-tails.tsx:42]: Make this description more obvious that this field is optional

File: packages/wrangler/src/tail/createTail.ts [hunk 2]
```diff
@@ -7,6 +7,8 @@ export { translateCLICommandToFilterMessage } from "./filters";
 export { jsonPrintLogs, prettyPrintLogs } from "./printing";
 import type { Request } from "undici";
 
+const TRACE_VERSION = "trace-v1";
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/tail/createTail.ts:10]: yay! 💯

File: packages/wrangler/src/pages/prompt-select-project.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { render, Text } from "ink";
+import SelectInput from "ink-select-input";
+import React from "react";
+import { listProjects } from "./projects";
+
+export async function promptSelectProject({
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/prompt-select-project.tsx:6]: 💯

File: packages/wrangler/src/pages/index.tsx [hunk 4]
```diff
@@ -81,6 +82,12 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
 							Publish.Options,
 							Publish.Handler
 						)
+						.command(
+							"tail [deployment-id]",
+							"Tail requests and logs on a deployment",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/index.tsx:87]: phrasing is a bit confusing here IMHO. Maybe smth like `Start a tailing session for a project's deployment and livestream logs from your Functions`. Doesn't have to be this wordy, but u get the gist

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 1 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/deployment-tails.tsx:42]: + maybe reword as `ID of the deployment to tail`

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+					"deployment ID is unknown.",
+			})
+			.options({
+				"project-name": {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:46]: is this the alternative if no deployment id arg is passed in?

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	const filters = translateCLICommandToFilterMessage({
+		header,
+		clientIp,
+		method,
+		samplingRate,
+		search,
+		status,
+	});
+
+	await metrics.sendMetricsEvent("begin pages log stream", {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:192]: `B`egin pages...?

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+	const onCloseTail = (() => {
+		let didTerminate = false;
+
+		return async () => {
+			if (didTerminate) return;
+
+			tail.terminate();
+			await deleteTail();
+			await metrics.sendMetricsEvent("end pages log stream", {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:218]: `E`nd pages ...?

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 8]
```diff
@@ -0,0 +1,772 @@
...
+		mockWebSockets.splice(0);
+		unsetAllMocks();
+	});
+
+	/**
+	 * Interaction with the tailing API, including tail creation,
+	 * deletion, and connection.
+	 */
+	describe("API interaction", () => {
+		it("should throw an error if deployment isn't provided", async () => {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:37]: if deployment `id`

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 9]
```diff
@@ -0,0 +1,772 @@
...
+			await expect(runWrangler("pages deployment tail")).rejects.toThrow();
+		});
+
+		it("creates and then delete tails by deployment ID", async () => {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:41]: delete`s`

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 10]
```diff
@@ -0,0 +1,772 @@
...
+
+		it("errors when the websocket closes unexpectedly", async () => {
+			const api = mockTailAPIs();
+			api.ws.close();
+
+			await expect(
+				runWrangler(
+					"pages deployment tail mock-deployment-id --project-name mock-project"
+				)
+			).rejects.toThrow();
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:79]: does this throw a specific error? if so we could test that as well

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 11]
```diff
@@ -0,0 +1,772 @@
...
+			        [31mX [41;31m[[41;97mERROR[41;31m][0m [1m  Error: { complex: 'error' }[0m
+
+			        "
+		      `);
+			expect(std.warn).toMatchInlineSnapshot(`""`);
+		});
+	});
+});
+
+/* helpers */
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:485]: some of these are identical to wrangler's `tail` test helpers. i wonder if we could extract and reuse 🤔 . Or could that potentially become a problem in the future?

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 12]
```diff
@@ -0,0 +1,772 @@
...
+			);
+			await expect(api.nextMessageJson()).resolves.toHaveProperty(
+				"debug",
+				true
+			);
+		});
+	});
+
+	describe("filtering", () => {
+		it("sends sampling rate filters", async () => {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:95]: if any of these throw specific errors, let's test for that as well

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 13]
```diff
@@ -0,0 +1,772 @@
...
+			const api = mockTailAPIs();
+			const search = "filterMe";
+
+			await runWrangler(
+				`pages deployment tail mock-deployment-id --project-name mock-project --search ${search}`
+			);
+			expect(api.requests.creation[0].body).toEqual(`[{"query":"${search}"}]`);
+		});
+
+		it("sends everything but the kitchen sink", async () => {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:205]: 😅

**Review by CarmenPopoviciu — CHANGES_REQUESTED** [review]

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts (hunk 11 above)
Comment (jrf0110) [reply packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:485]: I think for test helpers I'll keep them separated. However, for some of the duplicated controller code, I think we should unify

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts (hunk 10 above)
Comment (jrf0110) [reply packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:79]: Yeeeee I could specify some command output

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 1 above)
Comment (WalshyDev) [reply packages/wrangler/src/pages/deployment-tails.tsx:42]: We really need to make the deployment ID more visible too if we want to use the UUID. Right now you can only know it from the URL which isn't great, we should display it more visibly or also accept the hostname.

Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:42]: Good callout (to both comments). I do want to point out that you can run `wrangler pages deployment list` to get the UUID of your last (i think) 30 deployments, but your point about this being not great UX still stands. The happy path here is to omit the deployment ID altogether and let the tool get your _latest_ deployment for you.

@WalshyDev what do you think about renaming the positional argument `deployment-id` to `deployment` and then allow for the following values:

* Deployment ID
  - e.g. `wrangler pages deployment tail 2d5ca4e4-5701-11ed-9b6a-0242ac120002`
* Deployment URL
  - e.g. `wrangler pages deployment tail https://1ec5d691.example-project-power-ballad.pages.dev/`
* Deployment short code
  - e.g. `wrangler pages deployment tail 1ec5d691`

Comment (WalshyDev) [reply packages/wrangler/src/pages/deployment-tails.tsx:42]: I think that would be amazing!

File: packages/wrangler/src/pages/index.tsx [hunk 14]
```diff
@@ -81,6 +82,12 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
 							Publish.Options,
 							Publish.Handler
 						)
+						.command(
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/index.tsx:85]: can we make sure that `wrangler pages tail --help` shows the appropriate help message here? we're referencing this command in the docs

**Comment by CarmenPopoviciu** [comment]
@jrf0110  One more nit, that I've recently bumped into and I think could use some polishing 🙏 

Currently, if I run `wrangler pages deployment tail` the terminal output shows me smth along the lines of:
```
No deploymentId specified. Using latest deployment for  production environment.
Running tail on deploymentId: <some_deployment_id>
```
I find this somewhat confusing, because it always makes me wonder "is this thing streaming now or is it still attempting to connect?`. Would be helpful IMHO to change the wording here a bit, to express:
1. the attempt to create tail
2. whether attempt was successful or not
3. being ready to receive logs

`wrangler tail <worker_name>` outputs 👇 
```
Successfully created tail, expires at 2022-11-03T16:51:39Z
Connected to floral-mountain-6c7a, waiting for logs...
```

It doesn't have to be exactly that output, but smth more in that direction

File: packages/wrangler/src/tail/createTail.ts [hunk 15]
```diff
@@ -59,6 +61,58 @@ function makeDeleteTailUrl(
...
+	projectName,
+	deploymentId,
+	filters,
+	debug = false,
+}: CreatePagesTailOptions) {
+	const tailRecord = await fetchResult<TailCreationApiResponse>(
+		`/accounts/${accountId}/pages/projects/${projectName}/deployments/${deploymentId}/tails`,
+		{
+			method: "POST",
+			body: JSON.stringify(filters.filters),
```
Comment (GregBrimble) [inline packages/wrangler/src/tail/createTail.ts:83]:
```suggestion
			body: JSON.stringify(filters),
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 16]
```diff
@@ -0,0 +1,772 @@
...
+		});
+
+		it("sends search filters", async () => {
+			const api = mockTailAPIs();
+			const search = "filterMe";
+
+			await runWrangler(
+				`pages deployment tail mock-deployment-id --project-name mock-project --search ${search}`
+			);
+			expect(api.requests.creation[0].body).toEqual(`[{"query":"${search}"}]`);
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:202]:
```suggestion
			expect(api.requests.creation[0].body).toEqual(`{"filters":[{"query":"${search}"}]}`);
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 17]
```diff
@@ -0,0 +1,772 @@
...
+			const cliFilters =
+				`--sampling-rate ${sampling_rate} ` +
+				status.map((s) => `--status ${s} `).join("") +
+				method.map((m) => `--method ${m} `).join("") +
+				`--header ${header} ` +
+				client_ip.map((c) => `--ip ${c} `).join("") +
+				`--search ${query} ` +
+				`--debug`;
+
+			const expectedWebsocketMessage = `[{"sampling_rate":0.69},{"outcome":["ok","exception","exceededCpu","exceededMemory","unknown"]},{"method":["GET","POST","PUT"]},{"header":{"key":"X-HELLO","query":"world"}},{"client_ip":["192.0.2.1","self"]},{"query":"onlyTheseMessagesPlease"}]`;
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:223]:
```suggestion
			const expectedWebsocketMessage = `{"filters":[{"sampling_rate":0.69},{"outcome":["ok","exception","exceededCpu","exceededMemory","unknown"]},{"method":["GET","POST","PUT"]},{"header":{"key":"X-HELLO","query":"world"}},{"client_ip":["192.0.2.1","self"]},{"query":"onlyTheseMessagesPlease"}]}`;
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 18]
```diff
@@ -0,0 +1,772 @@
...
+
+		it("sends single IP filters", async () => {
+			const api = mockTailAPIs();
+			const fakeIp = "192.0.2.1";
+
+			await runWrangler(
+				`pages deployment tail mock-deployment-id --project-name mock-project --ip ${fakeIp}`
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"client_ip":["${fakeIp}"]}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:179]:
```suggestion
				`{"filters":[{"client_ip":["${fakeIp}"]}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 19]
```diff
@@ -0,0 +1,772 @@
...
+			);
+		});
+
+		it("sends header filters with a query", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --header X-CUSTOM-HEADER:some-value"
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"header":{"key":"X-CUSTOM-HEADER","query":"some-value"}}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:167]:
```suggestion
				`{"filters":[{"header":{"key":"X-CUSTOM-HEADER","query":"some-value"}}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 20]
```diff
@@ -0,0 +1,772 @@
...
+			);
+		});
+
+		it("sends header filters without a query", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --header X-CUSTOM-HEADER"
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"header":{"key":"X-CUSTOM-HEADER"}}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:157]:
```suggestion
				`{"filters":[{"header":{"key":"X-CUSTOM-HEADER"}}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 21]
```diff
@@ -0,0 +1,772 @@
...
+			expect(api.requests.creation[0].body).toEqual(`[{"method":["POST"]}]`);
+		});
+
+		it("sends multiple HTTP method filters", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --method POST --method GET"
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"method":["POST","GET"]}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:147]:
```suggestion
				`{"filters":[{"method":["POST","GET"]}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 22]
```diff
@@ -0,0 +1,772 @@
...
+				`[{"outcome":["exception","exceededCpu","exceededMemory","unknown","canceled"]}]`
+			);
+		});
+
+		it("sends single HTTP method filters", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --method POST"
+			);
+			expect(api.requests.creation[0].body).toEqual(`[{"method":["POST"]}]`);
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:138]:
```suggestion
			expect(api.requests.creation[0].body).toEqual(`{"filters":[{"method":["POST"]}]}`);
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 23]
```diff
@@ -0,0 +1,772 @@
...
+			);
+		});
+
+		it("sends multiple status filters", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --status error --status canceled"
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"outcome":["exception","exceededCpu","exceededMemory","unknown","canceled"]}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:129]:
```suggestion
				`{"filters":[{"outcome":["exception","exceededCpu","exceededMemory","unknown","canceled"]}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 24]
```diff
@@ -0,0 +1,772 @@
...
+			expect(api.requests.creation[0].body).toEqual(`[{"sampling_rate":0.25}]`);
+		});
+
+		it("sends single status filters", async () => {
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --status error"
+			);
+			expect(api.requests.creation[0].body).toEqual(
+				`[{"outcome":["exception","exceededCpu","exceededMemory","unknown"]}]`
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:119]:
```suggestion
				`{"filters":[{"outcome":["exception","exceededCpu","exceededMemory","unknown"]}]}`
```

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts [hunk 25]
```diff
@@ -0,0 +1,772 @@
...
+
+			const tooLow = runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --sampling-rate -5"
+			);
+			await expect(tooLow).rejects.toThrow();
+
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project --sampling-rate 0.25"
+			);
+			expect(api.requests.creation[0].body).toEqual(`[{"sampling_rate":0.25}]`);
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:110]:
```suggestion
			expect(api.requests.creation[0].body).toEqual(`{"filters":[{"sampling_rate":0.25}]}`);
```

File: packages/wrangler/src/tail/createTail.ts [hunk 26]
```diff
@@ -59,6 +61,58 @@ function makeDeleteTailUrl(
 		: `/accounts/${accountId}/workers/scripts/${workerName}/tails/${tailId}`;
 }
 
+interface CreatePagesTailOptions {
+	accountId: string;
+	projectName: string;
+	deploymentId: string;
+	filters: TailFilterMessage;
```
[the remark this answers is not in the document]
Comment (jrf0110) [reply packages/wrangler/src/tail/createTail.ts:68]: image

This is the payload that is also sent to Workers. It includes the array of filters within the envelope. It's not a great IMO, but I'm going to leave as is for now

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 6 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:192]: None of the other event names are capitalized so keeping as is

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 5 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:46]: project-name is actually always required, but we'll prompt you for it if not specified

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 1 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx]: wasn't exactly sure how the short code thingy worked, so I just did by url for now

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 27]
```diff
@@ -137,34 +138,54 @@ export async function Handler({
...
+				"Could not find deployment match url: " + deployment,
+				1
+			);
+		}
+
+		deploymentId = targetDeployment.id;
+	} else {
+		if (format === "pretty") {
+			logger.log(
+				"No deployment specified. Using latest deployment for ",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:185]: remove white space after `for`. It's not necessary 👇 . Or perhaps use `No deployment specified. Using latest deployment for ${environment} environment.`

Screenshot 2022-11-11 at 11 52 54

File: packages/wrangler/src/pages/index.tsx [hunk 28]
```diff
@@ -83,8 +83,9 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
 							Publish.Handler
 						)
 						.command(
-							"tail [deployment-id]",
-							"Tail requests and logs on a deployment",
+							"tail [deployment]",
+							"Start a tailing session for a project's deployment and " +
+								" livestream logs from your Functions",
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/index.tsx:88]: remove white space. There's already a trailing one on the line above 👇
Screenshot 2022-11-11 at 11 57 08

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 29]
```diff
@@ -137,34 +138,54 @@ export async function Handler({
...
-	if (!deploymentId && !projectName && isInteractive) {
+	// When not specifying a deployment, we need a project name to use the latest
+	// Or when specifying by deployment URL
+	if ((!deployment || isUrl(deployment)) && !projectName && isInteractive) {
 		projectName = await promptSelectProject({ accountId });
 	}
 
-	if (!deploymentId && !projectName) {
-		throw new FatalError("Must specify a project name or deployment ID.", 1);
+	if (!deployment && !projectName) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:151]: shouldn't this throw when `deployment && !projectName`? Running `wrangler pages deployment tail <deployment>` returns 👇 . IMHO we should throw here and not allow an invalid request to go through

Screenshot 2022-11-11 at 12 14 43

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 30]
```diff
@@ -137,34 +138,54 @@ export async function Handler({
...
 	);
 	const accountId = await requireAuth(pagesConfig);
+	let deploymentId = deployment;
 
 	projectName = projectName || pagesConfig.project_name;
 
-	if (!deploymentId && !projectName && isInteractive) {
+	// When not specifying a deployment, we need a project name to use the latest
+	// Or when specifying by deployment URL
+	if ((!deployment || isUrl(deployment)) && !projectName && isInteractive) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:147]: I wonder...is there any use case in which not specifying the project name shouldn't trigger this prompt?
The `tail`ing request is sent to `/account/<account_id>/pages/projects/<project_name>/deployments`, which I makes `projectName` mandatory, unless we can somehow deduct it based on the params being passed to the `tail` command?

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/deployment-tails.tsx:147]: noticing that somewhere before this code we set `projectName = projectName || pagesConfig.project_name`, which makes me think we assume? that if no project name is passed as an argument, 
```
const pagesConfig = getConfigCache<PagesConfigCache>(
		PAGES_CONFIG_CACHE_FILENAME
	);
	```
will always have the answers for us. But from what I'm seeing, that doesn't always seem to be the case 👇 

Screenshot 2022-11-11 at 13 01 41

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 31]
```diff
@@ -137,34 +138,54 @@ export async function Handler({
...
-			throw new FatalError("No deployments for environment: " + environment, 1);
+		if (!targetDeployment) {
+			throw new FatalError(
+				"Could not find deployment match url: " + deployment,
+				1
+			);
+		}
+
+		deploymentId = targetDeployment.id;
+	} else {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:182]: what about if deployment is a `deploymentID`? Currently, running `wrangler pages deployment tail <deploymentID>` behaves the same way as if I didn't pass any deployment at all 👇 

Screenshot 2022-11-11 at 13 24 48

File: packages/wrangler/src/pages/deployment-tails.tsx [hunk 32]
```diff
@@ -137,34 +138,54 @@ export async function Handler({
...
+	// Deployment is URL
+	if (deployment?.startsWith("https://")) {
+		const { hostname: deploymentHostname } = new URL(deployment);
+		const targetDeployment = envDeployments.find(
+			(d) => new URL(d.url).hostname === deploymentHostname
 		);
 
-		if (envDeployments.length === 0) {
-			throw new FatalError("No deployments for environment: " + environment, 1);
+		if (!targetDeployment) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/deployment-tails.tsx:174]: should we maybe do the same if we pass in an invalid/non-existing `deploymentId`? Currently there's no check for that and we simply fallback to the latest deployment. IMO we could perform the check and:

- keep the fallback but warn user the provided id is invalid and that we're falling back to latest
- throw and exit

Not sure what would make more sense 🤔 . Maybe first is less disruptive...

File: packages/wrangler/src/tail/createTail.ts [hunk 33]
```diff
@@ -69,6 +69,16 @@ interface CreatePagesTailOptions {
...
 
+/**
+ * Create and connect to a Pages Function Tail.
+ *
+ * Under the hood, this function
+ * - Registers a new Tail with the API
+ * - Connects to the tail worker
+ * - Sends any filters over the connection
+ *
+ * @returns a websocket connection, an expiration, and a function to call to delete the tail
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/tail/createTail.ts:80]: do we return the expiration?

Comment (jrf0110) [reply packages/wrangler/src/tail/createTail.ts:80]: Yup. On line 123

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 32 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:174]: In the case of no deploymentID provided, we fallback and warn the user, which I think is what we want. When a user provides an invalid deployment ID, this will fail by way of 404-ing at the `POST /tails` endpoint, thereby throwing and exiting. I think that's fairly ideal, right?

Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:174]: Jaykay, I just tried an invalid deployment I borked something. Fixing

Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx:166]: Okay _now_ it's working as described by my first comment :)

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 31 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx]: Yeahhhhh that was a regression. Fixed!!

File: packages/wrangler/src/pages/deployment-tails.tsx (hunk 30 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/deployment-tails.tsx]: I've cleaned this up. It's better to just not cache the project name at all. Only account ID

File: packages/wrangler/src/pages/index.tsx (hunk 28 above)
Comment (jrf0110) [reply packages/wrangler/src/pages/index.tsx:88]: Done

File: packages/wrangler/src/__tests__/pages-deployment-tail.test.ts (hunk 10 above)
Comment (jrf0110) [reply packages/wrangler/src/__tests__/pages-deployment-tail.test.ts:119]: Added output snapshot
